### DISCUSSION TITLE
GOLang: RDKIT integration.

### DISCUSSION THREAD

**Discussion by RamakrishnaChilaka** [rdkit/rdkit#4153] [Q&A]
Please confirm, If golang wrapper is available for RDKit ? Thanks.

**Comment by greglandrum** [comment]
The RDKit does not directly have a golang wrapper available, but the new CFFI library could certainly be used to create one:
https://greglandrum.github.io/rdkit-blog/technical/2021/05/01/rdkit-cffi-part1.html

**Reply by RamakrishnaChilaka** [reply]
@greglandrum , Thank a lot for exciting functionality of CFFI. However after building and trying to run the example program gives the following error.
**gcc -o demo -I $RDBASE/Code -I $CONDA_PREFIX/include -I /home/ramakrishna/rdkit/Code/MinimalLib -L**
```
$RDBASE/lib/librdkitcffi.so demo.c
/tmp/ccC2wanT.o: In function `canon_smiles':
demo.c:(.text+0x2d): undefined reference to `get_mol'
demo.c:(.text+0x49): undefined reference to `get_smiles'
/tmp/ccC2wanT.o: In function `main':
demo.c:(.text+0xa2): undefined reference to `enable_logging'
demo.c:(.text+0xac): undefined reference to `version'
collect2: error: ld returned 1 exit status
```

**Reply by RamakrishnaChilaka** [reply]
@greglandrum , Kindly, give your inputs. Thanks.

**Reply by greglandrum** [reply]
@RamakrishnaChilaka the blog post includes an example command line to build the demo program:
```
cc -o demo.exe -I $RDBASE/Code demo.c $RDBASE/lib/librdkitcffi.so
```
First you need to have either built librdkitcffi.so yourself or downloaded it following the instructions in the blog post.

Unfortunately this is still very much an intermediate/advanced developer feature since we haven't figured out how to make it accessible to people with less experience working with C/C++. I'm not going to be able to help you through the basics of building and linking C programs here.

**Comment by apahl** [comment]
Possible beginnings of a Go wrapper:

```Go
package main

// #include "cffiwrapper.h"
// #cgo CFLAGS: -I/home/pahl/anaconda3/envs/chem/include/rdkit -I/home/pahl/anaconda3/envs/chem/include
// #cgo LDFLAGS: -L/home/pahl/anaconda3/envs/chem/lib -lrdkitcffi
import "C"
import (
	"errors"
	"fmt"
)

type Mol struct {
	json string
	size C.ulong
}

func MolFromSmiles(smiles string) (Mol, error) {
	smilesPtr := C.CString(smiles)
	defer C.free_ptr(smilesPtr)
	var size C.ulong
	pklPtr := C.get_mol(smilesPtr, &size, nil)
	if pklPtr == nil {
		return Mol{}, errors.New("no Mol")
	}
	defer C.free_ptr(pklPtr)
	jsonPtr := C.get_json(pklPtr, size, nil)
	json := C.GoString(jsonPtr)
	defer C.free_ptr(jsonPtr)
	return Mol{json, size}, nil

}

func (m Mol) Descriptors() string {
	jsonPtr := C.CString(m.json)
	defer C.free_ptr(jsonPtr)
	molPtr := C.get_mol(jsonPtr, &m.size, nil)
	defer C.free_ptr(molPtr)
	descPtr := C.get_descriptors(molPtr, m.size)
	defer C.free_ptr(descPtr)
	result := C.GoString(descPtr)
	return result
}

func main() {
	smiles := "c1ccccc1C(=O)N"
	mol, err := MolFromSmiles(smiles)
	if err == nil {
		desc := mol.Descriptors()
		fmt.Println(desc)
	}
}
```

Compile and run with:  
```
$ LD_LIBRARY_PATH=/home/pahl/anaconda3/envs/chem/lib go run rdkit.go 
{"exactmw":121.05276,"amw":121.139,"lipinskiHBA":2.0,"lipinskiHBD":2.0,"NumRotatableBonds":1.0,"NumHBD":1.0,"NumHBA":1.0,"NumHeteroatoms":2.0,"NumAmideBonds":1.0,"FractionCSP3":0.0,"NumRings":1.0,"NumAromaticRings":1.0,"NumAliphaticRings":0.0,"NumSaturatedRings":0.0,"NumHeterocycles":0.0,"NumAromaticHeterocycles":0.0,"NumSaturatedHeterocycles":0.0,"NumAliphaticHeterocycles":0.0,"NumSpiroAtoms":0.0,"NumBridgeheadAtoms":0.0,"NumAtomStereoCenters":0.0,"NumUnspecifiedAtomStereoCenters":0.0,"labuteASA":53.29775,"tpsa":43.09,"CrippenClogP":0.7855,"CrippenMR":34.95989,"chi0v":4.87234,"chi1v":2.65348,"chi2v":1.0583,"chi3v":1.0583,"chi4v":0.60241,"chi0n":4.87234,"chi1n":2.65348,"chi2n":1.0583,"chi3n":1.0583,"chi4n":0.60241,"hallKierAlpha":-1.31,"kappa1":5.82003,"kappa2":2.30676,"kappa3":1.2877,"Phi":1.49171}
```

I am using an Anaconda install of the RDKit for this (in an environment called "chem") and have copied the compiled `librdkitcffi.so` to the `lib` folder that is used for the `LD_LIBRARY_PATH`.  
I also symlinked `librdkitcffi.so` to `librdkitcffi.so.1` in the same directory.
Put the header file `cffiwrapper.h` in the same directory as the Go file. I have modified it a bit, but I guess, the original file will also work. If not, let me know, then I can share my modified version as well.

This is of course all still very rough, I wrote this mainly, because I was also curious, if it would be possible to call the RDKit from Go as well.  
If you want to reproduce the code, you would need to adapt the absolute paths to your setup.  
If this turns out to be useful, it might make sense to put the code in a proper repository, parsing the various JSON outputs into proper objects, and so on. But I would need to do some more testing, first, I just wanted to put this here as a POC.

Some general remarks:  
It does not seem to be possible to store the binary pickle string of the molecule on the Go side, therefore the molecule is always converted to JSON.  
Rust was so far the only language where I managed to directly store the pickled representation.

Kind regards,  
Axel

**Reply by greglandrum** [reply]
> It does not seem to be possible to store the binary pickle string of the molecule on the Go side, therefore the molecule is always converted to JSON.
> Rust was so far the only language where I managed to directly store the pickled representation.

@apahl : at least some languages require you to store binary strings in a byte array. This is how things are handled in Python, for example. It looks like golang does have such a thing, so maybe that would help?

**Reply by apahl** [reply]
Yes, thanks, I tried that, but the type of the [Byte array]() is an `unsafe.Pointer` on the C side and that can not be passed to the C functions of the FFI wrapper (which expect a `*C.char` in Go Speak) anymore and I have not yet found out how to convert from `unsafe.Pointer` to `*C.char` (the other way round is possible).
There might be other options, but so far I could not find a solution.

**Reply by apahl** [reply]
Update:
This cast from `unsafe.Pointer` to `(*C.char)` in the `Descriptors` function seems to work 
(`pklPtr := (*C.char)(C.CBytes(m.pkl))`).
Thanks a lot for the pointer, Greg (pun intended):
```Go
package main

// #include "cffiwrapper.h"
// #cgo CFLAGS: -I/home/pahl/anaconda3/envs/chem/include/rdkit -I/home/pahl/anaconda3/envs/chem/include
// #cgo LDFLAGS: -L/home/pahl/anaconda3/envs/chem/lib -lrdkitcffi
import "C"
import (
	"errors"
	"unsafe"
)

type Mol struct {
	pkl  []byte
	size C.ulong
}

func MolFromSmiles(smiles string) (Mol, error) {
	smilesPtr := C.CString(smiles)
	defer C.free_ptr(smilesPtr)
	var size C.ulong
	pklPtr := C.get_mol(smilesPtr, &size, nil)
	if pklPtr == nil {
		return Mol{}, errors.New("no Mol")
	}
	defer C.free_ptr(pklPtr)
	pkl := C.GoBytes(unsafe.Pointer(pklPtr), C.int(size))
	return Mol{pkl, size}, nil
}

func (m Mol) Descriptors() string {
	pklPtr := (*C.char)(C.CBytes(m.pkl))
	defer C.free_ptr(pklPtr)
	descPtr := C.get_descriptors(pklPtr, m.size)
	defer C.free_ptr(descPtr)
	result := C.GoString(descPtr)
	return result
}

func main() {
	smiles := "c1ccccc1C(=O)N"
	mol, err := MolFromSmiles(smiles)
	if err == nil {
		desc := mol.Descriptors()
		fmt.Println(desc)
	}
}
```

**Reply by RamakrishnaChilaka** [reply]
@apahl , Thanks a lot for the detailed explanation. When i tried running, I am getting the following error. please review and help.
```
# command-line-arguments
/home/ramakrishna/anaconda3/envs/chem/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/ramakrishna/anaconda3/envs/chem/lib/librdkitcffi.so: undefined reference to `log@GLIBC_2.29'
/home/ramakrishna/anaconda3/envs/chem/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/ramakrishna/anaconda3/envs/chem/lib/librdkitcffi.so: undefined reference to `exp@GLIBC_2.29'
/home/ramakrishna/anaconda3/envs/chem/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/ramakrishna/anaconda3/envs/chem/lib/librdkitcffi.so: undefined reference to `pow@GLIBC_2.29'
collect2: error: ld returned 1 exit status
```

**Reply by apahl** [reply]
Without more information (e.g. details of the Linux OS, what the `command-line-arguments` is supposed to do), this is difficult to tell, but this looks to me like the versions of the `glibc` on your system (in the conda env) and the one the CFFI wrapper was compiled with, do not match ([stack overflow](https://stackoverflow.com/questions/57155333/undefined-reference-to-log2fglibc-2-27-undefined-reference-to-logfglibc-2-2)).  
You could try to install a different GCC in your conda env or to compile the CFFI wrappers from source on the machine where you want to run `command-line-arguments`.

KR Axel

**Reply by RamakrishnaChilaka** [reply]
@apahl, thanks for looking into this. Please find the image, which gives details about the error.
![image](https://user-images.githubusercontent.com/49393831/121669051-8eebfe00-cac9-11eb-9406-b3fec4173382.png)

**Reply by RamakrishnaChilaka** [reply]
@apahl , the error shows to use GLIB_2.29, but the most recent version of GLIBC available is 2.27 from Conda. Should i install glibc manually or compile librdkitcffi from source ? What do you think is the most efficient way to do this ? Thanks.

**Reply by apahl** [reply]
It is a bit strange that glibc 2.27 should be newest one on conda, but I can also not find a newer version.  
On my Ubuntu machine I use the same GCC as you in your conda env, but I have glibc 2.31 (can you check in your conda env with `ldd --version` what the actual version of the glibc is?).  
I think I would try compiling from source, but I have never done this myself, so I can not help with that.

**Reply by RamakrishnaChilaka** [reply]
@apahl , Please find the ldd --version response.
```
ldd (Ubuntu GLIBC 2.27-3ubuntu1.2) 2.27
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper
```

**Reply by RamakrishnaChilaka** [reply]
@apahl , I found the error, the max version of glibc that can be installed on ubuntu 18.04 is 2.27. I used a docker container with ubuntu version 20.04 with glibc version 2.31 and everything works as expected. Thanks for all the help. :)

**Comment by ptosco** [comment]
@RamakrishnaChilaka Did you try adding `-lm` to `LDFLAGS`:
```
#cgo LDFLAGS: -lm -L/home/pahl/anaconda3/envs/chem/lib -lrdkitcffi
```
Just guessing as I have never compiled a Go source, but to me it looks like it could simply need linking `libm`.

**Reply by RamakrishnaChilaka** [reply]
Yes @ptosco , But I did got the same error :), thanks for looking into this.
